Gateway misses to report failed request as metric

**Describe the bug**

During preparing for the ZEP https://github.com/zeebe-io/enhancements/pull/16 I did some experiments to find good metrics which could show us how we can improve fail over and what the current state is.

I run a benchmark with one partition and when the leader is restarted we can see that the processing etc drops, which is expected.

![general](https://user-images.githubusercontent.com/2758593/122369623-a4947400-cf5e-11eb-9568-b7b8e47d7a30.png)

We have metrics for the gateway which should show failure rates, but they are missing data points in this particular event, which is unexpected. To note we using standalone gateway.

![grpc-error](https://user-images.githubusercontent.com/2758593/122375566-9eed5d00-cf63-11eb-8562-2cc362351c88.png)


In the logs we can see that the requests are failing.

```
I 2021-06-16T14:01:11.557431017Z java.util.concurrent.ExecutionException: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: Time out between gateway and broker: Request type command-api-1 timed out in 15000 milliseconds
	at java.util.concurrent.CompletableFuture.reportGet(Unknown Source) ~[?:?]
	at java.util.concurrent.CompletableFuture.get(Unknown Source) ~[?:?]
	at io.zeebe.ResponseChecker.run(ResponseChecker.java:41) [app.jar:1.0.0-alpha7]
Caused by: io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: Time out between gateway and broker: Request type command-api-1 timed out in 15000 milliseconds
	at io.grpc.Status.asRuntimeException(Status.java:535) ~[app.jar:1.0.0-alpha7]
	at io.grpc.stub.ClientCalls$StreamObserverToCallListenerAdapter.onClose(ClientCalls.java:478) ~[app.jar:1.0.0-alpha7]
	at io.grpc.PartialForwardingClientCallListener.onClose(PartialForwardingClientCallListener.java:39) ~[app.jar:1.0.0-alpha7]
	at io.grpc.ForwardingClientCallListener.onClose(ForwardingClientCallListener.java:23) ~[app.jar:1.0.0-alpha7]
	at me.dinowernli.grpc.prometheus.MonitoringClientCallListener.onClose(MonitoringClientCallListener.java:50) ~[app.jar:1.0.0-alpha7]
	at io.grpc.internal.ClientCallImpl.closeObserver(ClientCallImpl.java:553) ~[app.jar:1.0.0-alpha7]
	at io.grpc.internal.ClientCallImpl.access$300(ClientCallImpl.java:68) ~[app.jar:1.0.0-alpha7]
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInternal(ClientCallImpl.java:739) ~[app.jar:1.0.0-alpha7]
	at io.grpc.internal.ClientCallImpl$ClientStreamListenerImpl$1StreamClosed.runInContext(ClientCallImpl.java:718) ~[app.jar:1.0.0-alpha7]
	at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37) ~[app.jar:1.0.0-alpha7]
	at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123) ~[app.jar:1.0.0-alpha7]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(Unknown Source) ~[?:?]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(Unknown Source) ~[?:?]
	at java.lang.Thread.run(Unknown Source) ~[?:?]
 
I 2021-06-16T14:01:11.567708552Z 14:01:11.567 [Thread-2] WARN  io.zeebe.ResponseChecker - Request failed
``` 

If we take a look at the grpc metrics, then we see the reason. Requests are rejected with RESOURCE_EX, which is ignored by our gateways as failure.

![reason](https://user-images.githubusercontent.com/2758593/122375663-b9bfd180-cf63-11eb-89c1-843e8fc7a4fa.png)

The problem is that the command api is turned on to early. Before the reprocessing of the stream processor is done. This is because the openFuture of the StreamProcessor is already completed [before the ReprocessingStateMachine is started](https://github.com/camunda-cloud/zeebe/blob/develop/engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/StreamProcessor.java#L123-L145).

```java
      healthCheckTick();
      openFuture.complete(null);

      final ReProcessingStateMachine reProcessingStateMachine =
          new ReProcessingStateMachine(processingContext);

      // disable writing to the log stream but for reprocessing checks
      processingContext.disableLogStreamWriter();
      processingContext.enableReprocessingStreamWriter();

      recoverFuture = reProcessingStateMachine.startRecover(snapshotPosition);

      actor.runOnCompletion(
          recoverFuture,
          (lastReprocessedPosition, throwable) -> {
            if (throwable != null) {
              LOG.error("Unexpected error on recovery happens.", throwable);
              onFailure(throwable);
            } else {
              onRecovered(lastReprocessedPosition);
              new StreamProcessorMetrics(partitionId)
                  .recoveryTime(System.currentTimeMillis() - startTime);
            }
          });
```

This behavior distorted the metrics, like backpressure but also failure rates etc.

<!-- A clear and concise description of what the bug is. -->

**To Reproduce**
<!--
Steps to reproduce the behavior

If possible add a minimal reproducer code sample
 - when using the Java client: https://github.com/zeebe-io/zeebe-test-template-java

-->
Run benchmark with one partition, delete the leader and check the metrics.

**Expected behavior**
<!-- A clear and concise description of what you expected to happen. -->
I would expect that the stream processor is only marked as open if it can work on new commands, which means after reprocessing/replay. We shoul'd only enable the command api for that partition if we are really able to handle incoming traffic.


**Environment:**
 - OS: k8
 - Zeebe Version: 1.0/snapshot/0..26/always
 - Configuration: 1 partition 3 replica

--- a/engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/StreamProcessor.java
+++ b/engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/StreamProcessor.java
@@ -120,7 +120,6 @@ public class StreamProcessor extends Actor implements HealthMonitorable {
           new ProcessingStateMachine(processingContext, this::shouldProcessNext);
 
       healthCheckTick();
-      openFuture.complete(null);
 
       final ReProcessingStateMachine reProcessingStateMachine =
           new ReProcessingStateMachine(processingContext);
@@ -302,6 +301,8 @@ public class StreamProcessor extends Actor implements HealthMonitorable {
     if (!shouldProcess) {
       setStateToPausedAndNotifyListeners();
     }
+
+    openFuture.complete(null);
   }
 
   private void onFailure(final Throwable throwable) {

--- a/engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/StreamProcessor.java
+++ b/engine/src/main/java/io/camunda/zeebe/engine/processing/streamprocessor/StreamProcessor.java
@@ -345,7 +345,7 @@ public class StreamProcessor extends Actor implements HealthMonitorable {
       return HealthStatus.UNHEALTHY;
     }
 
-    if (!processingStateMachine.isMakingProgress()) {
+    if (processingStateMachine == null || !processingStateMachine.isMakingProgress()) {
       return HealthStatus.UNHEALTHY;
     }
 

--- a/engine/src/test/java/io/camunda/zeebe/engine/processing/streamprocessor/StreamProcessorHealthTest.java
+++ b/engine/src/test/java/io/camunda/zeebe/engine/processing/streamprocessor/StreamProcessorHealthTest.java
@@ -178,7 +178,7 @@ public class StreamProcessorHealthTest {
 
   private StreamProcessor getErrorProneStreamProcessor() {
     streamProcessor =
-        streamProcessorRule.startTypedStreamProcessor(
+        streamProcessorRule.startTypedStreamProcessorNotAwaitOpening(
             processingContext -> {
               final MutableZeebeState zeebeState = processingContext.getZeebeState();
               mockedLogStreamWriter = new WrappedStreamWriter();

--- a/engine/src/test/java/io/camunda/zeebe/engine/processing/streamprocessor/StreamProcessorInconsistentPositionTest.java
+++ b/engine/src/test/java/io/camunda/zeebe/engine/processing/streamprocessor/StreamProcessorInconsistentPositionTest.java
@@ -100,7 +100,7 @@ public final class StreamProcessorInconsistentPositionTest {
     // when
     final TypedRecordProcessor typedRecordProcessor = mock(TypedRecordProcessor.class);
     final var streamProcessor =
-        firstStreamProcessorComposite.startTypedStreamProcessor(
+        firstStreamProcessorComposite.startTypedStreamProcessorNotAwaitOpening(
             (processors, context) ->
                 processors
                     .onEvent(ValueType.PROCESS_INSTANCE, ELEMENT_ACTIVATING, typedRecordProcessor)

--- a/engine/src/test/java/io/camunda/zeebe/engine/processing/streamprocessor/StreamProcessorTest.java
+++ b/engine/src/test/java/io/camunda/zeebe/engine/processing/streamprocessor/StreamProcessorTest.java
@@ -93,10 +93,11 @@ public final class StreamProcessorTest {
   public void shouldCallStreamProcessorLifecycleOnFail() throws InterruptedException {
     // given
     final CountDownLatch failedLatch = new CountDownLatch(1);
-    streamProcessorRule.startTypedStreamProcessor(
+    streamProcessorRule.startTypedStreamProcessorNotAwaitOpening(
         (processors, state) ->
             processors.withListener(
                 new StreamProcessorLifecycleAware() {
+
                   @Override
                   public void onRecovered(final ReadonlyProcessingContext context) {
                     throw new RuntimeException("force fail");

--- a/engine/src/test/java/io/camunda/zeebe/engine/util/StreamProcessingComposite.java
+++ b/engine/src/test/java/io/camunda/zeebe/engine/util/StreamProcessingComposite.java
@@ -53,15 +53,29 @@ public class StreamProcessingComposite {
   public StreamProcessor startTypedStreamProcessor(
       final StreamProcessorTestFactory factory, final Consumer<TypedRecord> onProcessedListener) {
     return startTypedStreamProcessor(
-        (processingContext) -> {
-          zeebeState = processingContext.getZeebeState();
-          lastProcessedPositionState = processingContext.getLastProcessedPositionState();
-          processingContext.onProcessedListener(onProcessedListener);
-          return factory.build(
-              TypedRecordProcessors.processors(
-                  zeebeState.getKeyGenerator(), processingContext.getWriters()),
-              processingContext);
-        });
+        (processingContext) ->
+            createTypedRecordProcessors(factory, onProcessedListener, processingContext));
+  }
+
+  public StreamProcessor startTypedStreamProcessorNotAwaitOpening(
+      final StreamProcessorTestFactory factory, final Consumer<TypedRecord> onProcessedListener) {
+    return startTypedStreamProcessorNotAwaitOpening(
+        (processingContext) ->
+            createTypedRecordProcessors(factory, onProcessedListener, processingContext));
+  }
+
+  private TypedRecordProcessors createTypedRecordProcessors(
+      final StreamProcessorTestFactory factory,
+      final Consumer<TypedRecord> onProcessedListener,
+      final io.camunda.zeebe.engine.processing.streamprocessor.ProcessingContext
+          processingContext) {
+    zeebeState = processingContext.getZeebeState();
+    lastProcessedPositionState = processingContext.getLastProcessedPositionState();
+    processingContext.onProcessedListener(onProcessedListener);
+    return factory.build(
+        TypedRecordProcessors.processors(
+            zeebeState.getKeyGenerator(), processingContext.getWriters()),
+        processingContext);
   }
 
   public StreamProcessor startTypedStreamProcessor(final TypedRecordProcessorFactory factory) {
@@ -71,6 +85,28 @@ public class StreamProcessingComposite {
   public StreamProcessor startTypedStreamProcessor(
       final int partitionId, final TypedRecordProcessorFactory factory) {
     return streams.startStreamProcessor(
+        getLogName(partitionId),
+        zeebeDbFactory,
+        (processingContext -> {
+          zeebeState = processingContext.getZeebeState();
+          lastProcessedPositionState = processingContext.getLastProcessedPositionState();
+          return factory.createProcessors(processingContext);
+        }));
+  }
+
+  public StreamProcessor startTypedStreamProcessorNotAwaitOpening(
+      final StreamProcessorTestFactory factory) {
+    return startTypedStreamProcessorNotAwaitOpening(factory, r -> {});
+  }
+
+  public StreamProcessor startTypedStreamProcessorNotAwaitOpening(
+      final TypedRecordProcessorFactory factory) {
+    return startTypedStreamProcessorNotAwaitOpening(partitionId, factory);
+  }
+
+  public StreamProcessor startTypedStreamProcessorNotAwaitOpening(
+      final int partitionId, final TypedRecordProcessorFactory factory) {
+    return streams.startStreamProcessorNotAwaitOpening(
         getLogName(partitionId),
         zeebeDbFactory,
         (processingContext -> {

--- a/engine/src/test/java/io/camunda/zeebe/engine/util/StreamProcessorRule.java
+++ b/engine/src/test/java/io/camunda/zeebe/engine/util/StreamProcessorRule.java
@@ -137,6 +137,17 @@ public final class StreamProcessorRule implements TestRule {
     return startTypedStreamProcessor(startPartitionId, factory);
   }
 
+  public StreamProcessor startTypedStreamProcessorNotAwaitOpening(
+      final StreamProcessorTestFactory factory) {
+    return streamProcessingComposite.startTypedStreamProcessorNotAwaitOpening(factory, r -> {});
+  }
+
+  public StreamProcessor startTypedStreamProcessorNotAwaitOpening(
+      final TypedRecordProcessorFactory factory) {
+    return streamProcessingComposite.startTypedStreamProcessorNotAwaitOpening(
+        startPartitionId, factory);
+  }
+
   public StreamProcessor startTypedStreamProcessor(
       final int partitionId, final TypedRecordProcessorFactory factory) {
     return streamProcessingComposite.startTypedStreamProcessor(partitionId, factory);

--- a/engine/src/test/java/io/camunda/zeebe/engine/util/TestStreams.java
+++ b/engine/src/test/java/io/camunda/zeebe/engine/util/TestStreams.java
@@ -217,13 +217,22 @@ public final class TestStreams {
       final ZeebeDbFactory zeebeDbFactory,
       final TypedRecordProcessorFactory typedRecordProcessorFactory) {
     final SynchronousLogStream stream = getLogStream(log);
-    return buildStreamProcessor(stream, zeebeDbFactory, typedRecordProcessorFactory);
+    return buildStreamProcessor(stream, zeebeDbFactory, typedRecordProcessorFactory, true);
+  }
+
+  public StreamProcessor startStreamProcessorNotAwaitOpening(
+      final String log,
+      final ZeebeDbFactory zeebeDbFactory,
+      final TypedRecordProcessorFactory typedRecordProcessorFactory) {
+    final SynchronousLogStream stream = getLogStream(log);
+    return buildStreamProcessor(stream, zeebeDbFactory, typedRecordProcessorFactory, false);
   }
 
   private StreamProcessor buildStreamProcessor(
       final SynchronousLogStream stream,
       final ZeebeDbFactory zeebeDbFactory,
-      final TypedRecordProcessorFactory factory) {
+      final TypedRecordProcessorFactory factory,
+      final boolean awaitOpening) {
     final var storage = createRuntimeFolder(stream);
     final var snapshot = storage.getParent().resolve(SNAPSHOT_FOLDER);
 
@@ -245,7 +254,11 @@ public final class TestStreams {
             .streamProcessorFactory(factory)
             .eventApplierFactory(eventApplierFactory)
             .build();
-    streamProcessor.openAsync(false).join(15, TimeUnit.SECONDS);
+    final var openFuture = streamProcessor.openAsync(false);
+
+    if (awaitOpening) {
+      openFuture.join(15, TimeUnit.SECONDS);
+    }
 
     final LogContext context = logContextMap.get(logName);
     final ProcessorContext processorContext =

